core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java

```diff
@@ -617,14 +610,14 @@ public synchronized CatchupFuture catchup(Map<String, Long> journalSequenceNumbe
 
   @Override
   public synchronized void checkpoint() throws IOException {
-    // TODO(feng): consider removing this once we can automatically propagate
-    //             snapshots from standby master
     try (RaftJournalAppender client = new RaftJournalAppender(mServer, this::createClient,
-        mRawClientId)) {
-      mSnapshotAllowed.set(true);
+        mRawClientId); RaftClient raftClient = createClient()) {
       catchUp(mStateMachine, client);
-      mStateMachine.takeLocalSnapshot();
-      // TODO(feng): maybe prune logs after snapshot
+      mStateMachine.setAllowLeaderSnapshots(true);
+      // taking a manual checkpoint can take a long time, users are warned about this, so we set
+      // a long timeout for the operation
+      RaftClientReply reply = raftClient.getSnapshotManagementApi().create(Integer.MAX_VALUE);
+      processReply(reply, "failed to take checkpoint");
```

Comment:
I wonder if there are cases where this could block forever and would bring down the system since the state lock is owned while this is called?

And the danger with putting a timeout here is that if the operation is cancelled, the snapshot may still be going on, but then the state lock is released so it could take an invalid snapshot.

I think the only safe way to do this is to put all this logic and the state lock inside the take snapshot function? what do you think?

Reply:
I've moved the `StateLockManager` down into the `checkpoint` function call and I pass it to the state machine. Let me know what you think.